xmake/rules/c++/modules/modules_support/dependency_scanner.lua

```diff
@@ -593,145 +447,104 @@ function fallback_generate_dependencies(target, jsonfile, sourcefile, preprocess
+-- get source modulefile for external target deps
+function get_targetdeps_modules(target)
+    local sourcefiles
+    for _, dep in ipairs(target:orderdeps()) do
+        local sourcebatch = dep:sourcebatches()["c++.build.modules.builder"]
+        if sourcebatch and sourcebatch.sourcefiles then
+            for _, sourcefile in ipairs(sourcebatch.sourcefiles) do
+                local fileconfig = dep:fileconfig(sourcefile)
+                local public = (fileconfig and fileconfig.public and not fileconfig.external) or false
+                if public then
+                    sourcefiles = sourcefiles or {}
+                    table.insert(sourcefiles, sourcefile)
+                    target:fileconfig_add(sourcefile, {external = true})
                 end
             end
         end
     end
+    return sourcefiles
 end
 
-function get_modulehash(target, modulepath)
-    local key = path.directory(modulepath) .. target:name()
-    return hash.uuid(key):split("-", {plain = true})[1]:lower()
-end
+-- cull unused packages modules
+-- removed named module not used in the translation units
+-- when building a library we only cull external modules because we need module objectfiles to be linked inside the library
+-- on an executable we cull explicitly referenced module
+function cull_unused_modules(target, modules)
 
-function get_metafile(target, modulefile)
-    local outputdir = get_outputdir(target, modulefile)
-    return path.join(outputdir, path.filename(modulefile) .. ".meta-info")
-end
+    local cull_all_modules = target:kind() == "executable"
+
+    local needed_modules = {}
+    for _, module in pairs(modules) do
+        local fileconfig = target:fileconfig(module.sourcefile)
+        local external = fileconfig and fileconfig.external
+        if not (cull_all_modules and external) then
+            goto CONTINUE
+        end
 
-function get_outputdir(target, module)
-    local cachedir = modules_cachedir(target)
-    local modulehash = get_modulehash(target, module.path or module)
-    local outputdir = path.join(cachedir, modulehash)
-    if not os.exists(outputdir) then
-        os.mkdir(outputdir)
+        if module.provides and module.requires then
+            table.join2(needed_modules, _fill_needed_module(target, modules, module))
+        end
+
+        ::CONTINUE::
+    end
+
+    local culled = {}
+    for objectfile, module in pairs(modules) do
+        -- if cull_all_modules and modules.provides then
+        --     local name,_,_ = compiler_support.get_provided_module(module)
+        --     if module.requires then
+        --         for required, _ in pairs(module.requires) do
+        --             table.insert(needed_modules, required)
+        --         end
+        --     end
+        --     if table.find(needed_modules, name) then
+        --        culled[objectfile] = module
+        --     end
+        -- else
+             culled[objectfile] = module
```

Comment:
cull_unused_modules just return the original modules. if it does not work, we should remove it.

or 

``lua
function cull_unused_modules(target, modules)
    return modules
```

@Arthapz